### PR TITLE
Add JSON parsing logic

### PR DESCRIPTION
**Pull request by xichen2020**
cc @notbdu 

This PR adds JSON parsing logic to parse an incoming string or a `[]byte` into a value, with options to support various pools and maximum parsing depth.

### CODE REVIEW

File: parser/json/parser_benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+func BenchmarkParseRawString(b *testing.B) {
+	for _, s := range []string{`""`, `"a"`, `"abcd"`, `"abcdefghijk"`, `"qwertyuiopasdfghjklzxcvb"`} {
+		b.Run(s, func(b *testing.B) {
+			benchmarkParseRawString(b, s)
+		})
+	}
+}
+
+func benchmarkParseRawString(b *testing.B, s string) {
+	b.ReportAllocs()
```
Comment (notbdu) [inline parser/json/parser_benchmark_test.go]: Oh cool, didn't know about this bench fn.

File: parser/json/parser_benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+}
+
+// NB(xichen): This should work in theory but have not been stress tested.
+// This is currently only called in tests.
+func s2b(s string) []byte {
+	if len(s) == 0 {
+		return nil
+	}
+
+	// NB(xichen): we need to declare a real byte slice so internally the compiler
```
Comment (notbdu) [inline parser/json/parser_benchmark_test.go]: Ah, this is interesting. Thanks for the note.

File: parser/json/parser_benchmark_test.go [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+	// knows to use an unsafe.Pointer to keep track of the underlying memory so tha
+	// once the slice's array pointer is updated with the pointer to the string's
+	// underlying bytes, the compiler won't prematurely GC the memory when the string
+	// goes out of scope.
+	var b []byte
+	byteHeader := (*reflect.SliceHeader)(unsafe.Pointer(&b))
```
Comment (notbdu) [inline parser/json/parser_benchmark_test.go]: So does casting to `reflect.TYPEHeader` give you access to the golang runtime's internal representation of a bye slice and a string?

Comment (xichen2020) [reply parser/json/parser_benchmark_test.go]: Yes.

**Pull request merged**